test/test_broker.hpp

```diff
@@ -1052,22 +1104,17 @@ class test_broker {
 
                     e.con.reset();
 
-                    auto const& sei_opt = e.session_expiry_interval;
-                    if (sei_opt && sei_opt.value() !=
-                        std::chrono::seconds(MQTT_NS::session_never_expire)) {
-                        e.tim_session_expiry = std::make_shared<as::steady_timer>(ioc_);
-                        e.tim_session_expiry->expires_after(sei_opt.value());
+                    if (e.session_expiry_interval && e.session_expiry_interval.value() != std::chrono::seconds(MQTT_NS::session_never_expire)) {
+                        e.tim_session_expiry = std::make_shared<as::steady_timer>(ioc_, e.session_expiry_interval.value());
                         e.tim_session_expiry->async_wait(
-                            [&, wp = std::weak_ptr<as::steady_timer>(e.tim_session_expiry)]
-                            (MQTT_NS::error_code ec) {
-                                auto sp = wp.lock();
+                            [&, wp = std::weak_ptr<as::steady_timer>(e.tim_session_expiry)](MQTT_NS::error_code ec) {
                                 if (!ec) {
-                                    auto& idx = sessions_.get<tag_tim>();
-                                    idx.erase(sp);
+                                    sessions_.get<tag_tim>().erase(wp.lock());
```

Comment:
I think that checking wp lifetime first is better.
It's not introduced by your PR. The original code is also not good.

`erase(nullptr);` doesn't match anything in our design so works expectedly but I don't like this kinds of design.


Please use the following pattern.

```cpp
if (auto sp = wp.lock()) { // first check wp and acquire sp
    if (!ec) { // timer fired, not cancelled
        // do erase operation
    }
}
```